data-gen/img-processing/downloadAndGetHostedUrls.js

```diff
@@ -0,0 +1,164 @@
+};
+
+/**
+ * Downloads numToDownload unique avatar images to the file system (path ./imgs)
+ * @param {Integer} numToDownload
+ */
+const downloadProfileImgs = async (numToDownload = 100) => {
+  console.log(`Downloading ${numToDownload} randomly generated user avatars to ./img...`);
+  if (numToDownload > 100) {
+    throw new Error('Please limit number of images to 100 at a time or less to respect rate limits.');
+  }
+
+  let eyesOptions = [ 1, 2, 3, 4, 5, 6, 7, 9, 10 ];
+  let noseOptions = [ 2, 3, 4, 5, 6, 7, 8, 9 ];
+  let mouthOptions = [ 1, 3, 5, 6, 7, 9, 10, 11 ];
+  let eLen = eyesOptions.length;
+  let nLen = noseOptions.length;
+  let mLen = mouthOptions.length;
+  let basePath = 'https://api.adorable.io/avatars/face/';
+
+  // Map urls to fetch promises that resolve to true or reject to false
+
+  let downloadPromises = new Array(numToDownload).fill(0).map((_, idx) => {
+    let hexColor = getRandomHexColor();
+    let url = basePath
+      + `eyes${eyesOptions[getRandIdx(eLen)]}/`
+      + `nose${noseOptions[getRandIdx(nLen)]}/`
+      + `mouth${mouthOptions[getRandIdx(mLen)]}/`
+      + hexColor;
+
+    return fetch(url)
+      .then(res => res.body)
+      .then(readable => {
+        let writePath = path.resolve(__dirname, 'imgs', `${hexColor}${idx.toString().padStart(3, '0')}.png`);
+        let writeable = fs.createWriteStream(writePath);
+        readable.on('end', () => {
+          writeable.end();
+        });
+        readable.pipe(writeable);
+        return true;
+      })
+      .catch(err => {
+        console.error(err);
+        return false;
+      });
+  });
+
+  // For now, return await is unnecessary, but if this fn is being in anything
+  // other than a CLI one-time-use script, then a return might be expected & will save time.
+  return await Promise.all(downloadPromises);
+};
+
+cloudinary.config({
+  cloud_name: process.env.CLOUD_NAME,
+  api_key: process.env.API_KEY,
+  api_secret: process.env.API_SECRET
+});
+
+/**
+ * Promisifies cloudinary's resources_by_tag function
+ * @returns {Promise}
+ */
+const getResourcesByTag = (tag, options) => {
+  return new Promise((resolve, reject) => {
+    cloudinary.api.resources_by_tag(tag, options, (err, res) => {
+      if (err) {
+        reject(err);
+      }
+
+      resolve(res);
+    });
+  });
+};
+
+/**
+ * Gets of list of URLs of resources uploaded to Cloudinary
+ */
+const getHostedUrls = async () => {
+  let badgeUrls;
+  let avatarUrls = [];
+
+  // Get 16 badge URLs
+  try {
+    let results = await getResourcesByTag('badge', { max_results: 16 });
+    if (results.resources.length) {
+      badgeUrls = results.resources.map(resource => resource.secure_url);
+    }
+  } catch (e) {
+    console.error(e);
+    return;
+  }
+
+  // Get 750 user profile (avatar) URLs
+  let nextCursor;
+  for (let i = 0; i < 8; i++) {
+    let options = {
+      max_results: 100
+    };
+
+    if (nextCursor) {
+      options.next_cursor = nextCursor;
```

Comment:
What is "next_cursor"? I only see it here and line 147

Reply:
Sure! This is part of the JSON response from Cloudinary's API, which allows pagination, since each request is limited to maximum 100 urls in a response. So I record the next_cursor string after each response, and feed it into the next request to pick up where I left off